[Woo POS][Product Search] Local search algorithm update

samiuelson opened the pull request:
<!-- Remember about a good descriptive title. -->
WOOPRD-517

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR changes product search functionality by adding support for searching by SKU, description, and short description fields. The implementation:
1. Adds version checking to determine if WooCommerce Core supports the enhanced search capability (v9.9.0 and above)
2. For stores with WooCommerce v9.9.0+:
* Maintains compatibility with modernized backend search – mplements a tokenized search that breaks the query into terms
* Searches across product SKU and  name~, description, and short description fields~

3. For stores with older WooCommerce versions:
* Maintains backward compatibility with the backend (name, description, and short description fields)
4. Adds caching for the WooCommerce version check to avoid repeated checks

### Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

### Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->
#### Use WC beta tester to run 9.9.0 version:
Verify local search matches behavior of remote search for various search phrases:
* containing SKU
* containing product description
* containing short description
* containing multiple words
#### On WC version 9.8.5 
* Verify improved compatibility across local search algorithm and remote – simple search in the product.name field.

### The tests that have been performed
<!-- To give the reviewer an idea of what could be missed in terms of testing -->
Above

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->
N/A

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

samiuelson committed:
Implement local name or sku search

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

samiuelson committed:
Merge branch 'trunk' into wooprd-517-woo-posproduct-search-add-sku-in-the-local-filtering

samiuelson committed:
Adapt local search algorithm to WC version

samiuelson committed:
Add tests

samiuelson committed:
Satisfy detekt's complaints

samiuelson committed:
Update RELEASE-NOTES.txt

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/search/WooPosProductSearchPredicate.kt:
```diff
@@ -1,22 +1,48 @@
 
-    operator fun invoke(query: String): (Product) -> Boolean {
-        if (query.isBlank()) return { true }
+    operator fun invoke(query: String): (Product) -> Boolean =
```
np:

It feels that we could avoid negativity check here

```
    operator fun invoke(query: String): (Product) -> Boolean =
        when {
            query.isBlank() -> { _ -> true }
            isWooCoreSupportsNameOrSkuSearch() -> tokenizedSkuOrNameSearchPredicate(query)
            else -> simpleSearchPredicate(query)
        }
```

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/search/WooPosProductSearchPredicate.kt:
```diff
@@ -1,22 +1,48 @@
+    private fun isWooCoreSupportsNameOrSkuSearch(): Boolean {
+        cachedSupportsNameOrSkuSearch?.let { return it }
+        val wooCoreVersion = getWooCoreVersion() ?: return false
+        return wooCoreVersion >= WC_VERSION_SUPPORTS_NAME_OR_SKU_SEARCH
```
@samiuelson 

This is actually correct. You never assign `cachedSupportsNameOrSkuSearch`

samiuelson replied:
Fixed: edecb99

kidinov reviewed (commented):
@samiuelson 

I think for the versions below 9.9.0 they filter not only by the name. Notice that we don't find anything with "x" request while they do


https://github.com/user-attachments/assets/3fbb7b5d-fceb-4285-a44c-c2b425269570

For the search with SKU it still a bit off. Notice that the remote brings another value even after it supposed to be cached. It may be sorting involved here too

https://github.com/user-attachments/assets/a5a8cc30-18f0-438d-baf1-e0f3819d7c74

samiuelson committed:
Merge branch 'trunk' into wooprd-517-woo-posproduct-search-add-sku-in-the-local-filtering

samiuelson committed:
Update simpleSearchPredicate to match WP_Query logic

The 'search' parameter value is sanitized and used to build a SQL WHERE clause.
Only published posts (or posts the current user is allowed to see) are searched.
The search checks the post_title, post_content, and post_excerpt fields for matches.
By default, all search terms must be present somewhere in the post (AND logic).

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/search/WooPosProductSearchPredicate.kt:
```diff
@@ -1,22 +1,60 @@
+    private fun isWooCoreSupportsNameOrSkuSearch(): Boolean {
+        cachedSupportsNameOrSkuSearch?.let { return it }
+        val wooCoreVersion = getWooCoreVersion() ?: return false
+        return wooCoreVersion >= WC_VERSION_SUPPORTS_NAME_OR_SKU_SEARCH
```
I just figured out that you compare strings here. Probably we should use `isGreaterThanPluginVersion`

kidinov replied:
oops, actually no. We should use `semverCompareTo`

e.g

`val isSupported = version.semverCompareTo(AUTO_DRAFT_SUPPORTED_VERSION) >= 0`

samiuelson replied:
Good catch! Fixed: 48ccfb0

samiuelson replied:
...and: 742e38d

kidinov replied:
@samiuelson you still use `isGreaterThanPluginVersion`, not `semverCompareTo` to specific (9.9.0) version

samiuelson replied:
Thanks, fixed: 92f5f05

samiuelson committed:
Update simpleSearchPredicate to match WP_Query logic

The 'search' parameter value is sanitized and used to build a SQL WHERE clause.
Only published posts (or posts the current user is allowed to see) are searched.
The search checks the post_title, post_content, and post_excerpt fields for matches.
By default, all search terms must be present somewhere in the post (AND logic).

samiuelson committed:
Update tokenizedSkuOrNameSearchPredicate to match sku_or_name search logic

samiuelson committed:
Clean up code

samiuelson committed:
Clean up code

samiuelson committed:
Fix woo core version value caching

samiuelson committed:
Fix woo core version value caching

samiuelson committed:
Fix version comparing

samiuelson committed:
Fix version comparing

samiuelson committed:
Clean up code

samiuelson committed:
Clean up code

samiuelson committed:
Fix sem ver comparing

samiuelson committed:
Fix sem ver comparing

samiuelson committed:
Update tests

samiuelson committed:
Update tests

samiuelson commented:
Thanks for review 🙇 

I've rechecked the search logic with the help of GH Copilot. This is what I understand:
1. WC version < 9.9.0
> ### Fields Searched:
> The `search` parameter is passed to WP_Query, which (by default) searches the following WordPress post fields for products:
> - post_title (Product Title)
> - post_content (Product Description)
> - post_excerpt (Product Short Description)
> 
> These fields are standard for the 'product' post type in WooCommerce, as products are stored as custom post types in WordPress.
> 
> ### Search Logic
> The search term is sanitized and split into individual words.
> Each word must be present in at least one of the searched fields (title, content, excerpt). **By default, all search terms must match (AND logic between terms)**.
> The search is case-insensitive and matches partial words.
> Only published products (or products visible to the user) are returned.
> The search does not include meta fields (like SKU, custom attributes, etc.) unless a plugin or custom code extends the search logic.

2. WC version >= 9.9.0
> If the search_name_or_sku parameter is present, the code:
> * Tokenizes the query.
> * For each token, checks for a partial match in the product name (post_title) or SKU.
> * Supersedes (unsets) any other search parameters (search, search_sku, sku).
> If there is no match in name or SKU for a product, that product is not included in the results. **The code does not perform a fallback to a broader search (such as searching in description or shortDescription) if no matches are found.**

I've tested with the scenarios from your screen recordings, and I think it works now as expected.

Commits: 96d6859, 99eb8a2c3ae9fbd91bad7a86753919b8cc663e90

Please, let me know what you think, @kidinov !

samiuelson committed:
Merge branch 'trunk' into wooprd-517-woo-posproduct-search-add-sku-in-the-local-filtering

samiuelson committed:
Merge remote-tracking branch 'origin/wooprd-517-woo-posproduct-search-add-sku-in-the-local-filtering' into wooprd-517-woo-posproduct-search-add-sku-in-the-local-filtering

samiuelson commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/search/WooPosProductSearchPredicate.kt:
```diff
@@ -1,22 +1,65 @@
-            searchTerms.all { term ->
-                product.name.lowercase().contains(term)
+            tokens.all { token ->
+                product.name.contains(token, ignoreCase = true) || product.sku.contains(token, ignoreCase = true)
```
👎

samiuelson committed:
Satisfy detekt's complaints

kidinov reviewed (approved):
LGTM! Tests well too!

Just make sure the version comparison works correctly. The current implementation will break if, for example, version 9.86 is released without the sku search feature.

samiuelson committed:
Improve sem ver comparison

samiuelson committed:
Satisfy detekt's complaints
